pageserver/src/tenant/timeline.rs

```diff
@@ -3530,6 +3556,52 @@ impl Timeline {
                     "keeping {} because it is the latest layer",
                     l.filename().file_name()
                 );
+                // Collect delta key ranges that need image layers to allow garbage
+                // collecting the layers.
+                // It is not so obvious whether we need to propagate information only about
+                // delta layers. Image layers can form "stairs" preventing old image from been deleted.
+                // But image layers are in any case less sparse than delta layers. Also we need some
+                // protection from replacing recent image layers with new one after each GC iteration.
+                if l.is_incremental() {
+                    let mut to_remove: Vec<Key> = Vec::new();
+                    let mut insert_new_range = true;
+                    let start = l.get_key_range().start;
+                    let mut end = l.get_key_range().end;
+
+                    if !LayerMap::is_l0(&*l) {
+                        // check if this range overlaps with exited ranges
+                        let mut iter = wanted_image_layers.range_mut(..end);
+                        while let Some((prev_start, prev_end)) = iter.next_back() {
+                            if *prev_end >= start {
+                                // two ranges overlap
+                                if *prev_start <= start {
+                                    // combine with prev range
+                                    insert_new_range = false;
+                                    if *prev_end < end {
+                                        // extend prev range
+                                        *prev_end = end;
+                                        info!("Extend wanted image {}..{}", *prev_start, end);
+                                    }
+                                    break;
+                                } else {
+                                    to_remove.push(*prev_start);
+                                    if *prev_end > end {
+                                        end = *prev_end;
+                                    }
+                                }
+                            } else {
+                                break;
+                            }
+                        }
+                        for key in to_remove {
+                            wanted_image_layers.remove(&key);
+                        }
+                        if insert_new_range {
+                            info!("Wanted image {}..{}", start, end);
+                            wanted_image_layers.insert(start, end);
+                        }
+                    }
+                }
```

Comment:
From my previous comment about `LayerMap::is_l0`:

```suggestion
                if l.is_incremental() && !LayerMap::is_l0(&*l) {
                    let mut to_remove: Vec<Key> = Vec::new();
                    let mut insert_new_range = true;
                    let start = l.get_key_range().start;
                    let mut end = l.get_key_range().end;

                    // check if this range overlaps with exited ranges
                    let mut iter = wanted_image_layers.range_mut(..end);
                    while let Some((prev_start, prev_end)) = iter.next_back() {
                        if *prev_end >= start {
                            // two ranges overlap
                            if *prev_start <= start {
                                // combine with prev range
                                insert_new_range = false;
                                if *prev_end < end {
                                    // extend prev range
                                    *prev_end = end;
                                    info!("Extend wanted image {}..{}", *prev_start, end);
                                }
                                break;
                            } else {
                                to_remove.push(*prev_start);
                                if *prev_end > end {
                                    end = *prev_end;
                                }
                            }
                        } else {
                            break;
                        }
                    }
                    for key in to_remove {
                        wanted_image_layers.remove(&key);
                    }
                    if insert_new_range {
                        info!("Wanted image {}..{}", start, end);
                        wanted_image_layers.insert(start, end);
                    }
                }
```

I am surprised clippy didn't suggest this already.

This wins one indentation level. Of course it could be further refined into a function which name could tell us something, I think the args would be something like `wanted_image_layers: &mut _, layer: &L` so it might actually make sense.

Reply:
Also going throught the comments, we don't really need a comment like: `// extend prev range` because there's already a log message saying the exact same thing.

Noted a typo in comment:

```patch
-// check if this range overlaps with exited ranges
+// check if this range overlaps with existing ranges
```

Making these changes to my suggestion.